README.md
<p align="center">
  <img src="https://github.com/eclipse/upm/blob/master/docs/icons/upm_logo.png" height="150px" width="auto" algt="UPM Logo"/>
</p>

Eclipse UPM Sensor and Actuator Repository
==============

The Eclipse UPM repository provides software drivers for a wide variety of
commonly used sensors and actuators. These software drivers interact with the
underlying hardware platform (or microcontroller), as well as with the attached
sensors, through calls to [Eclipse MRAA](https://github.com/eclipse/mraa) APIs.

Programmers can access the interfaces for each sensor by including the sensor's
corresponding header file and instantiating the associated sensor class. In the
typical use case, a constructor initializes the sensor based on parameters that
identify the sensor, the I/O protocol used and the pin location of the sensor.
As of UPM 2.0, sensor initialization can also be done, in most cases, via
overloaded constructors that accept string identifiers.

We endorse additions that implement the generic C and C++ interfaces provided
with the libraries. With the 2.0 release, UPM introduces the following sensor
interfaces:
```
iAcceleration, iAngle, iButton, iClock, iCollision, iDistance,
iDistanceInterrupter, iEC, iElectromagnet, iEmg, iGas, iGps, iGyroscope,
iHallEffect, iHeartRate, iHumidity, iLight, iLineFinder, iMagnetometer,
iMoisture, iMotion, iOrp, iPH, iPressure, iProximity, iTemperature, iVDiv,
iWater.
```
The developer community is invited to propose new interfaces for actuator types.

The UPM project is an Eclipse IoT project.
You can read more about this [here](https://projects.eclipse.org/proposals/eclipse-upm).

### Example

A sensor/actuator is expected to work as such (here is the MMA7660 accelerometer API):
```C++
  // Instantiate an MMA7660 on I2C bus 0
  upm::MMA7660 *accel = new upm::MMA7660(MMA7660_DEFAULT_I2C_BUS,
                                         MMA7660_DEFAULT_I2C_ADDR);

  // place device in standby mode so we can write registers
  accel->setModeStandby();

  // enable 64 samples per second
  accel->setSampleRate(MMA7660_AUTOSLEEP_64);

  // place device into active mode
  accel->setModeActive();

  while (shouldRun)
    {
      float ax, ay, az;

      accel->getAcceleration(&ax, &ay, &az);
      cout << "Acceleration: x = " << ax
           << "g y = " << ay
           << "g z = " << az
           << "g" << endl;

      usleep(500000);
    }
```

Browse through the list of all [examples](https://github.com/eclipse/upm/tree/master/examples).

Multi-sensor samples for starter and specialized kits can be found in the
[iot-devkit-samples](https://github.com/intel-iot-devkit/iot-devkit-samples) repository.

### Supported Sensors

Supported [sensor list](http://iotdk.intel.com/docs/master/upm/modules.html) from API documentation.

### IDE Support and More

The UPM project includes support for multiple industrial-grade sensors, actuators, radios,
protocols and standards in use today. It is also highly integrated with the Eclipse IDE
version shipped with the Intel® oneAPI IoT Toolkit.
[Learn more](https://software.intel.com/content/www/us/en/develop/tools/oneapi/iot-toolkit.html).

### Installing UPM

Find notes on how to install UPM on various OS'es on this [page](docs/installing.md).

### Building UPM

See building documentation [here](docs/building.md).

[![Build Status](https://travis-ci.org/intel-iot-devkit/upm.svg?branch=master)](https://travis-ci.org/intel-iot-devkit/upm)
[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=upm-master&metric=alert_status)](https://sonarcloud.io/dashboard?id=upm-master)

### Guidelines and rules for new UPM contributions

Before you begin development, take a look at our naming [conventions](docs/naming.md).
The name you pick for a newly added sensor needs to be unique in the UPM library.

Next, review the project's [contribution guide](docs/contributions.md).

Make sure you add yourself as an author on every new code file submitted.
If you are providing a fix with significant changes, feel free to add yourself
as a contributor. Signing-off your commits and accepting the ECA is mandatory
for making new contributions to this project.

Documenting your code is also a big part of the task. We have a strict set of
tags used to classify our sensors and their capabilities. You can find out more
about this in our [section](docs/documentation.md) on documenting a sensor API.

Finally, if you really want to ensure consistency with the rest of the library,
and the intel-iot-devkit repositories in general, take a look at our extensive
[author guide](docs/guidelines.md).

API Documentation
==============

<a href="http://iotdk.intel.com/docs/master/upm"><img src="docs/icons/c++.png"/></a>
<a href="http://iotdk.intel.com/docs/master/upm/java"><img src="docs/icons/java.png"/></a>
<a href="http://iotdk.intel.com/docs/master/upm/python"><img src="docs/icons/python.png"/></a>
<a href="http://iotdk.intel.com/docs/master/upm/node"><img src="docs/icons/node.png"/></a>

### API Compatibility
Even if we try our best not to, every once in a while we are forced to modify
our API in a way that will break backwards compatibility. If you find yourself
unable to compile code that was working fine before a library update, make sure
you check the [API changes](docs/apichanges.md) section first.

### Changelog
Version changelog [here](docs/changelog.md).

### Known Limitations
List of known limitations [here](docs/knownlimitations.md).


docs/changelog.md
Changelog                         {#changelog}
===============

Here's a list summarizing some of the key undergoing changes to our library
from earlier versions:

### v2.0.1

 * Revised documentation to reflect project transfer to the Eclipse Foundation
 * Removed Python 2 bindings, old node.js versions from CI, disabled node.js
 builds by default
 * Updated and extended installation instructions for some Linux distributions

### v2.0.0

 * Reworked existing interfaces and extended the list to include iAcceleration,
 iAngle, iButton, iClock, iCollision, iDistance, iDistanceInterrupter, iEC,
 iElectromagnet, iEmg, iGas, iGps, iGyroscope, iHallEffect, iHeartRate, iHumidity,
 iLight, iLineFinder, iMagnetometer, iMoisture, iMotion, iOrp, iPH, iPressure,
 iProximity, iTemperature, iVDiv, iWater
 * Provided overloaded string based constructors for a good number of sensors,
 allowing initialization from external frameworks without type knowledge
 * Removed deprecated basic grove classes from project
 * Replaced 6 unsafe occurrences of sprintf() usage
 * Cleaned-up build system around target arch detection and cpack usage, added
 detection and resolution of tinyb library when used as a dependency
 * New sensors: 2jciebu01_ble, 2jciebu01_usb

### v1.7.1

 * Bumped library compatibility to new MRAA 2.0 APIs
 * Added installation details for Fedora
 * Minor fixes based on static analysis vulnerability report

### v1.7.0

 * Added Cordova bindings based on Java packages
 * Complete rework of the nmea_gps library giving more control over data
 acquisition and parsing
 * Added GTest support for unit tests
 * Removed calls to deprecated MRAA fast GPIO functions
 * Several bug fixes and improvements around SWIG type maps and gcc8 support
 * New sensors: kx122, kxtj3, bh1792, bh1749

### v1.6.0

 * Extended LED library to support the new MRAA gpio-leds APIs
 * Many CMake changes around SWIG wrapper generation and improved FindNodejs
 detection module
 * Several code fixes based on static code analysis
 * Improved documentation generation and Travis CI builds
 * Cleaned-up doxygen tags in headers and class names in JSON library files
 * New sensor: lis3dh

### v1.5.0

 * Introduced a flexible JSON format for technical sensor specifications, notes
 and classification. This is also used by our [new UPM website](http://upm.mraa.io)
 * Revised all C++ sensor examples and switched to stack allocations where
 possible along with other code and formatting clean-up
 * Significantly improved docker workflow, CI integration, sanity and
 consistency tests, static code scans and documentation generation
 * Several improvements to a couple of existing sensor drivers and better
 compatibility with subplatforms
 * Added new std::vector to AbstractList<> typemap for Java bindings and
 examples
 * New sensors: lis2ds12, lsm6ds3h, lsm6dsl, lidarlitev3

### v1.3.0

 * Finalized all required build system and JAVA binding changes to release the
 UPM libraries for Android Things targets
 * Enhanced the modules for lsm303 and tmp006 to support newer variants of the
 chips
 * Fixed Node.js detection on Ubuntu when installed with apt using official
 packages
 * Minor documentation changes
 * New sensors: lsm303agr, lsm303d, veml6070, tca9548a, rn2903

### v1.2.0

 * Improved JAVA binding compiler compatibility and added JAVA interfaces that
 match existing C++ interfaces
 * Ported Bosch drivers to C thus now they can be used on supported MCUs
 * Completely redesigned C/C++ example handling by cmake to avoid maintaining
 a list by hand
 * Fixed several existing drivers, mostly based on static code analysis reports
 * Added a sensor driver template plus bash script that can be easily used by
 developers to start writing their own sensor driver
 * Numerous documentation improvements
 * New sensors: p9813, abp, rsc, mmc35240, tcs37727, tmp006, mma8x6x, mag3110,
 hdc1000

### v1.1.0

 * Reworked cmake handling of internal and external dependencies
 * Documentation improvements for sensor names, sensor descriptions, external
 contributor guides and added new section on installing UPM on a supported OS
 * Fixed issues with C11 builds on some WRLinux gateways and Debian systems
 * New examples and functionality for speaker and ads1x15 modules
 * Modified JAVA builds to ensure the SWIG generated wrappers include all
 functionality and typemaps
 * New sensors: mcp2515, max30100, uartat, le910

### v1.0.2

 * Minor cmake changes for Java builds and Java examples are now toggled with
 the same BUILDEXAMPLES cmake switch
 * Made some improvements in the utilities class for our C drivers
 * Fixed issue with some drivers not being usable on subplatforms due to pin
 numbers using uint8 type
 * New sensors: ims, ecezo, mb704x, rf22

### v1.0.1

 * Warnings as errors enabled for C/CXX (fixed warnings from -Wall and others)
 * Fix for C/CXX compile flag append issue #485
 * C sensor libraries for: my9221, otp538u, guvas12d, ppd42ns
 * New sensor: ms5803

### v1.0.0

 * Added approximately 50 C sources for UPM drivers that can be used on both
 Unix and RTOS boards
 * Provided generic alternatives to the Grove set of classes as they will be
 deprecated over time and eventually removed
 * UPM modules are now being packed and published on NPM as sources rather than
 binaries to enhance compatibility with various boards and systems
 * Build both python2 and python3 modules if the corresponding python libs are
 available and removed cmake BUILDPYTHON3 switch
 * Converted python examples to run on both python2 and python3
 * Added 'upm' hierarchy to python module install path which changes the way
 UPM modules are imported
    Example:
        old: import pyupm_dfrph
        new: from upm import pyupm_dfrph
             or
             import upm.pyupm_dfrph
 * Added ctests for consistancy checking UPM repo as well as sanity checking
 for python2/3 modules and examples
 * Various documentation and driver fixes throughout

tests/check_pythonload.py
#!/usr/bin/python
#
# Copyright (c) 2017-2019 Intel Corporation
#
# This program and the accompanying materials are made available under the
# terms of the The MIT License which is available at
# https://opensource.org/licenses/MIT.
#
# SPDX-License-Identifier: MIT
#
import unittest
import imp
import glob
import os, sys

# Skip individual modules based on module name.  For example,
# pyupm_ozw will skip pyupm_ozw.py.
blacklist = [
             # OpenZwave-dependent modules
             'pyupm_ozw', 'aeotecdw2e', 'tzemt400', 'aeotecdsb09104',
             'aeotecss6', 'aeotecsdg2', 'ozwdump',
             # Requires mraa python module
             'curieimu',
             # Requires modbus
             'h803x', 'hwxpxx', 't3311',
             # Requires bacnet
             'e50hx', 'tb7300', 't8100',
             # Requires PIL
             'make_oled_pic'
             ]

class loadModule(unittest.TestCase):
    ''' The loadModule class loads all module which match a search string
    provided via argv.  If any of the target modules fails loading, this
    class will assert and provide a list of failing modules.'''
    def test_load_module(self):
        # Python version provided on the command line
        py_search_str = '../build/src/*/python3/*.py'
        if len(self.argv) > 0:
           py_search_str = self.argv[0]

        # Make sure the search string has a full path
        full_py_search_str = os.path.realpath(py_search_str)

        # Get all python modules matching full_py_search_str
        pyfiles = glob.glob(full_py_search_str)

        # Fail if no modules to test
        self.assertNotEqual(len(pyfiles), 0,
                'Failed to find any %s modules in %s' % \
                (py_search_str, full_py_search_str));

        # Test load each module
        failures = {}
        for pyfile in pyfiles:
            module = os.path.splitext(os.path.basename(pyfile))[0]

            # Don't load blacklisted modules
            if module in blacklist:
                print('Skipping blacklisted %s ...' % pyfile)
                continue

            # Append the path of this python module
            sys.path.append(os.path.dirname(pyfile))

            try:
                pyfile = imp.load_source(module, pyfile)
            except Exception as x:
                failures[pyfile] = x

        skeys = list(failures.keys())
        skeys.sort()
        self.assertEqual(len(failures), 0,
                '\n\n%s' % '\n'.join((['%s=%s' % (k, os.environ[k]) for k in list(os.environ.keys())])) +
                '\npython %s\n' % ' '.join(sys.version.strip().split()) +
                '\nFailed to load %d modules:\n' % len(failures) +
                '\n'.join(['%s: %s' % (k, failures[k]) for k in skeys]))

if __name__ == '__main__':
    # Allow passing from argv
    loadModule.argv = []
    for arg in sys.argv[1:]:
        loadModule.argv.append(sys.argv.pop())
    unittest.main()


tests/check_file_encoding.py
#!/usr/bin/python
#
# Copyright (c) 2017-2019 Intel Corporation
#
# This program and the accompanying materials are made available under the
# terms of the The MIT License which is available at
# https://opensource.org/licenses/MIT.
#
# SPDX-License-Identifier: MIT
#

import unittest
import os
import chardet

target_exts = ['.h', '.hpp', '.hxx', '.txt']
valid_encodings = ['ascii', 'utf-8']

class EncodingTests(unittest.TestCase):
    '''Non-ascii/utf-8 encodings can cause failures on downstream tools
    such as documentation-generation and python module loading.  This
    class helps find those files which could cause an encoding problem'''

    def test_headers_ascii(self):
        '''Assert/print list of:
            file:linenumber offending line
        for all lines of matching files which are not in valid_encodings'''
        # Keep a map of files with alternate encodes to report
        invalid_files = {}

        # Recusively search cwd for files with target_exts
        for root, dirs, files in os.walk(os.curdir):
            # Work on full paths
            for file in files:
                file = os.path.join(root, file)

                # Skip any files not ending with target_exts
                if not any(file.lower().endswith(x) for x in target_exts):
                    continue

                # Check each with chardet
                with open(file, 'r') as f:
                    for ndx, line in enumerate(f):
                        result = chardet.detect(line)
                        if not result['encoding'] in valid_encodings:
                            if not invalid_files.has_key(file):
                                invalid_files[file] = []
                            invalid_files[file].append([ndx,line])

        # Sort the failures by filename
        skeys = list(invalid_files.keys())
        skeys.sort()
        invalid_lines = ''
        for fn in skeys:
            for line in invalid_files[fn]:
                invalid_lines += '%s:%d %s' % (fn, line[0], line[1])

        self.assertEqual( len(invalid_files), 0,
                "\nThe following modules have alternate encodings:\n" + \
                invalid_lines)

if __name__ == '__main__':
    unittest.main()


src/ultrasonic/ultrasonic.cxx
/*
 * Author: Jun Kato <i@junkato.jp>
 * Copyright (c) 2015 Jun Kato.
 *
 * Thanks to Seeed Studio for a working arduino sketch
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#include <functional>
#include <iostream>
#include <stdlib.h>
#include <unistd.h>

#include "ultrasonic.hpp"

using namespace upm;

UltraSonic::UltraSonic(int pin)
{
    m_name = "UltraSonic";

    mraa_init();

    // setup pin
    m_pinCtx = mraa_gpio_init(pin);
    if (m_pinCtx == NULL) {
        fprintf(stderr, "Are you sure that pin%d you requested is valid on your platform?", pin);
        exit(1);
    }
    mraa_gpio_isr(m_pinCtx, MRAA_GPIO_EDGE_BOTH, &signalISR, this);
}

UltraSonic::~UltraSonic()
{
    // close pin
    mraa_gpio_isr_exit(m_pinCtx);
    mraa_gpio_close(m_pinCtx);
}

float
UltraSonic::getDistance()
{
    // output trigger signal
    mraa_gpio_dir(m_pinCtx, MRAA_GPIO_OUT);
    mraa_gpio_write(m_pinCtx, LOW);
    usleep(2);
    mraa_gpio_write(m_pinCtx, HIGH);
    usleep(5);
    mraa_gpio_write(m_pinCtx, LOW);

    // wait for the pulse,
    m_doWork = true;
    m_InterruptCounter = 0;
    mraa_gpio_dir(m_pinCtx, MRAA_GPIO_IN);

    // though do not wait over 25 [ms].
    int timer = 0;
    while (m_doWork && timer++ < 5) {
        // in 25 [ms], sound travels 25000 / 29 / 2 = 431 [cm],
        // which is more than 400 [cm], the max distance measurable with this sensor.
        usleep(5 * 1000); // 5 [ms]
    }

    // calc diff
    long diff = m_FallingTimeStamp.tv_usec - m_RisingTimeStamp.tv_usec;
    diff += (m_FallingTimeStamp.tv_sec - m_RisingTimeStamp.tv_sec) * 1000000;
    return timer >= 5 ? 0 : diff;
}

void
UltraSonic::signalISR(void* ctx)
{
    upm::UltraSonic* This = (upm::UltraSonic*) ctx;
    This->ackEdgeDetected();
}

void
UltraSonic::ackEdgeDetected()
{
    if (++m_InterruptCounter % 2 == 0) {
        gettimeofday(&m_FallingTimeStamp, NULL);
        m_doWork = false;
    } else {
        gettimeofday(&m_RisingTimeStamp, NULL);
    }
}


src/groveultrasonic/groveultrasonic.cxx
/*
 * Author: Jun Kato <i@junkato.jp>
 * Copyright (c) 2015 Jun Kato.
 *
 * Thanks to Seeed Studio for a working arduino sketch
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#include <functional>
#include <iostream>
#include <stdlib.h>
#include <unistd.h>

#include "groveultrasonic.hpp"

using namespace upm;

GroveUltraSonic::GroveUltraSonic(int pin)
{
    m_name = "GroveUltraSonic";

    mraa_init();

    // setup pin
    m_pinCtx = mraa_gpio_init(pin);
    if (m_pinCtx == NULL) {
        fprintf(stderr, "Are you sure that pin%d you requested is valid on your platform?", pin);
        exit(1);
    }
    mraa_gpio_isr(m_pinCtx, MRAA_GPIO_EDGE_BOTH, &signalISR, this);
}

GroveUltraSonic::~GroveUltraSonic()
{
    // close pin
    mraa_gpio_isr_exit(m_pinCtx);
    mraa_gpio_close(m_pinCtx);
}

float
GroveUltraSonic::getDistance()
{
    // output trigger signal
    mraa_gpio_dir(m_pinCtx, MRAA_GPIO_OUT);
    mraa_gpio_write(m_pinCtx, LOW);
    usleep(2);
    mraa_gpio_write(m_pinCtx, HIGH);
    usleep(5);
    mraa_gpio_write(m_pinCtx, LOW);

    // wait for the pulse,
    m_doWork = true;
    m_InterruptCounter = 0;
    mraa_gpio_dir(m_pinCtx, MRAA_GPIO_IN);

    // though do not wait over 25 [ms].
    int timer = 0;
    while (m_doWork && timer++ < 5) {
        // in 25 [ms], sound travels 25000 / 29 / 2 = 431 [cm],
        // which is more than 400 [cm], the max distance measurable with this sensor.
        usleep(5 * 1000); // 5 [ms]
    }

    // calc diff
    long diff = m_FallingTimeStamp.tv_usec - m_RisingTimeStamp.tv_usec;
    diff += (m_FallingTimeStamp.tv_sec - m_RisingTimeStamp.tv_sec) * 1000000;
    return timer >= 5 ? 0 : diff;
}

void
GroveUltraSonic::signalISR(void* ctx)
{
    upm::GroveUltraSonic* This = (upm::GroveUltraSonic*) ctx;
    This->ackEdgeDetected();
}

void
GroveUltraSonic::ackEdgeDetected()
{
    if (++m_InterruptCounter % 2 == 0) {
        gettimeofday(&m_FallingTimeStamp, NULL);
        m_doWork = false;
    } else {
        gettimeofday(&m_RisingTimeStamp, NULL);
    }
}
